### PR TITLE
Fixes for Android Unit Tests

### PR DESCRIPTION
**Pull request by amzn-phist**
## What does this PR do?

Fixes some issues that were preventing Android Unit Tests from being run.
Fixes an issue with dynamic modules on Android.  When the module name is changed to a real filename via adding the `AZ_TRAIT_OS_DYNAMIC_LIBRARY_PREFIX` and `AZ_TRAIT_OS_DYNAMIC_LIBRARY_EXTENSION`, write this change back to the `DynamicModuleHandle`.

## How was this PR tested?

Running `AzTestRunner` on various modules (AzCore, FastNoise, GradientSignal, Terrain, ...).  Some modules still have test errors, others now pass, but they can all run their tests now.

### CODE REVIEW

File: Code/Framework/AzCore/Tests/AZStd/Hashed.cpp [hunk 1]
```diff
@@ -1420,7 +1420,11 @@ namespace UnitTest
     TYPED_TEST_CASE(HashedSetDifferentAllocatorFixture, SetTemplateConfigs);
 
 #if GTEST_HAS_DEATH_TEST
+#if AZ_TRAIT_DISABLE_FAILED_DEATH_TESTS
+    TYPED_TEST(HashedSetDifferentAllocatorFixture, DISABLED_InsertNodeHandleWithDifferentAllocatorsLogsTraceMessages)
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/Tests/AZStd/Hashed.cpp:1424]: Do you know what the issue with the Google Death test are?
Is it not correctly causing a non-zero exit code?

**Review by lumberyard-employee-dm — APPROVED** [review]

File: Code/Framework/AzCore/Tests/AZStd/Hashed.cpp (hunk 1 above)
Comment (amzn-phist) [reply Code/Framework/AzCore/Tests/AZStd/Hashed.cpp:1424]: I'll have to run them again, but encountering one would prevent running the remaining tests.

Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/Tests/AZStd/Hashed.cpp:1424]: Ok then it is probably an issue with the Android runtime then, where it maybe having issues forking processes in an isolated manner..

**Review by cgalvan — APPROVED** [review]

File: Code/Framework/AzCore/Tests/AZStd/Hashed.cpp (hunk 1 above)
Comment (amzn-phist) [reply Code/Framework/AzCore/Tests/AZStd/Hashed.cpp:1424]: It does log a warning about forking process with threads, but the test itself is aborting while operating on a null `FILE*`.  Needs some investigation.

**Review by moraaar — COMMENTED** [review]
This couldn't have come in a better moment. Thanks @amzn-phist !

**Pull request merged**